### PR TITLE
Add a cancel button to the payment collection modal

### LINKED ISSUE #6646: [Mobile Payments] Add a cancel button to the payment collection modal
Image

> **CPP Inconsistencies**
>
> When collecting a payment, iOS shows a Cancel button, while Android does not.
> When collecting a payment, show a Cancel button on Android.
>
> – p91TBi-8P0-p2#cpp-inconsistencies

- [x] Add a cancel button to the collecting payment CPP modal

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #6646

### Description

Adds cancel button to all intermediate state of the payment flow

### Testing instructions

Run payment flow and click the cancel button on all intermediate states. Also try the same with interac refund

### Images/gif

https://user-images.githubusercontent.com/4923871/171850211-8f83de03-d735-4a3c-a5d9-7ae49d5518fc.mp4

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/payment/CardReaderPaymentViewState.kt [hunk 1]
```diff
@@ -60,33 +62,42 @@ sealed class ViewState(
...
     }
 
-    data class ProcessingPaymentState(override val amountWithCurrencyLabel: String) :
-        ViewState(
-            hintLabel = R.string.card_reader_payment_processing_payment_hint,
-            headerLabel = R.string.card_reader_payment_processing_payment_header,
-            paymentStateLabel = R.string.card_reader_payment_processing_payment_state,
-            illustration = R.drawable.img_card_reader_available
-        ),
+    data class ProcessingPaymentState(
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/payment/CardReaderPaymentViewState.kt]: ❓ I was thinking if we need a "Cancel" button in the `Capture` state? If the merchant cancels a payment flow in the `capturing` state, the payment status will still be shown as completed in the app as well as the wp-admin.
I know that pressing the back at a similar stage in payment flow has the same effect. However, adding an explicit "cancel" button and still showing the payment status as completed for the order might be confusing. WDYT?

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/CardReaderPaymentViewModelTest.kt [hunk 2]
```diff
@@ -919,7 +919,7 @@ class CardReaderPaymentViewModelTest : BaseUnitTest() {
         }
 
     @Test
-    fun `when loading data, then only progress is visible`() = testBlocking {
+    fun `when loading data, then only progress and cancel button is visible`() = testBlocking {
```
Comment (AnirudhBhat) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/CardReaderPaymentViewModelTest.kt]: ❓ What do you think about verifying that the cancel event is tracked correctly when the cancel button is tapped during various payment states? Also, tests that verify that the `Exit` event is being triggered when the cancel button is tapped during various states?

**Comment by AnirudhBhat** [comment]
Thank you for working on this @kidinov. I've tested this and the payment flow shows the "Cancel" button on all different payment flows as expected. However, the "Cancel" button is not visible on different refund flows for Interac. Is this expected?

Also, I saw strange behavior/bug during the simple payment flow. Steps below:
1. Start a simple payment flow
2. While the payment flow is in a processing state, tap the cancel button (tap cancel 2-3 seconds after the state has moved to processing state)
3. The simple flow operation gets canceled and the order doesn't get created in the app.
4. Even though no order gets created and the transaction was canceled while processing, Wisepad 3 device was showing the transaction as approved in its display and the flow in the card reader finished as if the transaction finished successfully.
5. The wp-admin shows the transaction and shows this order as successful. I think this is not expected? WDYT?


https://user-images.githubusercontent.com/1331230/172087179-d2306dbd-8c61-490b-821b-6b8b467195bd.mp4

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/payment/CardReaderPaymentViewState.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/payment/CardReaderPaymentViewState.kt]: I had the same thought and now you also think alike. Reverted cancel on the capturing step - 0fc76cafe9975e3a6e90630e7a2e2aa0969e8eb5

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/CardReaderPaymentViewModelTest.kt (hunk 2 above)
Comment (kidinov) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/CardReaderPaymentViewModelTest.kt]: Good point. Added some extra tests - 1afdd6497cd115e208238f5d44728844db5bd393

**Comment by kidinov** [comment]
@AnirudhBhat thank you for the review!

I completely forgot to add the buttons to the refund state 🤦‍♂️. Good catch!   Done 3390e9079b91209b89f43118e978c9db4ca5f3f0 and e600d44275714fd9c0eee888c5415e6ff64e740a

The bug you caught sounds like might need attention from @nbradbury . Adding a cancel button doesn't change much, android's back button is here anyway

**Comment by nbradbury** [comment]
@AnirudhBhat That's definitely not expected behavior, thanks for reporting it. I've filed [this issue](https://github.com/woocommerce/woocommerce-android/issues/6705) for it.

**Pull request merged**